### PR TITLE
FINERACT-2090: Restructure Loan creation and modification

### PR DESCRIPTION
**Pull request by adamsaghy**
Added further refactor that shared by @kjozsa! Thank you for it! ;)

The submit / modify loan application feature implementation contains various validations and conditions to decide on what logic to be executed based on the provided configuration.

The validations and conditions got mixed  a little bit and now it is hard to understand what is happening and why.

The job here was to refactor a little bit an d group the validations and logics to be executed together and extract the business logic into methods to separate them a little bit.

The rule of thumb was
- Validations
      - Data validations
      - Domain rule validations (sometimes it is only after entity was assembled)
- Assembling (from the json command → entity / data objects)
      - Assemble the entity
      - Some actions on the entity to be extracted into utility classes or service classes
      - Move some of the logic from service into here if make sense
      - Update product related details
      - New assemblers
          - collateral
          - charges
          - etc
- Actions on the loans
      - Generate schedule
      - Notes
      - GLIM
      - Calendar

## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanDisbursementDetailsAssembler.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    private final FromJsonHelper fromApiJsonHelper;
+
+    public List<LoanDisbursementDetails> fetchDisbursementData(final JsonObject command) {
+        final Locale locale = this.fromApiJsonHelper.extractLocaleParameter(command);
+        final String dateFormat = this.fromApiJsonHelper.extractDateFormatParameter(command);
+        List<LoanDisbursementDetails> disbursementDatas = new ArrayList<>();
+        if (command.has(LoanApiConstants.disbursementDataParameterName)) {
+            final JsonArray disbursementDataArray = command.getAsJsonArray(LoanApiConstants.disbursementDataParameterName);
+            if (disbursementDataArray != null && !disbursementDataArray.isEmpty()) {
+                int i = 0;
```
Comment (kjozsa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanDisbursementDetailsAssembler.java]: @adamsaghy I made an attempt to improve this code slightly further. 

Please check my commit at https://github.com/apache/fineract/commit/d544a35b6ae0172808bc63c396402bd4dcdd07ae and tell me if you wish to have a PR of this version.

Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanDisbursementDetailsAssembler.java]: Yes, LGTM. I am happy to add your changes as part of this PR.

**Comment by adamsaghy** [comment]
@taskain7 @vidakovic @kjozsa @ruchiD @alberto-art3ch Please as your time lets you give me a review. I know this is a monster PR, but it was unfortunately required to have a clearer restructure...

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/command/LoanChargeCommand.java [hunk 2]
```diff
@@ -18,44 +18,26 @@
...
 import java.time.LocalDate;
+import lombok.Data;
 import org.apache.fineract.portfolio.loanaccount.domain.LoanCharge;
 
 /**
  * Java object representation of {@link LoanCharge} API JSON.
  */
-public class LoanChargeCommand implements Comparable<LoanChargeCommand> {
+@Data
+public class LoanChargeCommand implements Serializable {
```
Comment (vidakovic) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/command/LoanChargeCommand.java:33]: ... minor: technically needs this

```
@Serial
private static final long serialVersionUID = 1L;
```

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/mapper/LoanChargeMapper.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Mapping(target = "id", source = "loanCharge.id")
+    @Mapping(target = "chargeId", source = "loanCharge.charge.id")
+    @Mapping(target = "amount", expression = "java(loanCharge.getAmount(currency).getAmount())")
+    @Mapping(target = "chargeTimeType", source = "loanCharge.chargeTimeType.value")
+    @Mapping(target = "chargeCalculationType", source = "loanCharge.chargeCalculation.value")
+    @Mapping(target = "dueDate", source = "loanCharge.dueDate")
+    LoanChargeCommand map(LoanCharge loanCharge, MonetaryCurrency currency);
+
+    default Set<LoanChargeCommand> map(Set<LoanCharge> loanCharge, MonetaryCurrency currency) {
```
Comment (vidakovic) [inline fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/mapper/LoanChargeMapper.java:41]: Great! Mapstruct... I think you could get away without a default function implementation and just provide:

```
Set<LoanChargeCommand> map(Set<LoanCharge> loanCharge, MonetaryCurrency currency);
```

I think Mapstruct is smart enough to create automatically a for loop and use the single value implementation on each element; with just the parameter "loanCharge"  I'd be 100% sure, with the additional currency parameter I'd say 90%.

Minor issue.

**Review by vidakovic — APPROVED** [review]
LGTM

File: fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/mapper/LoanChargeMapper.java (hunk 3 above)
Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/mapper/LoanChargeMapper.java:41]: I was getting the following exception:
```
error: The return type Set<LoanChargeCommand> is an abstract class or interface. Provide a non abstract / non interface result type or a factory method.
     Set<LoanChargeCommand> map(Set<LoanCharge> loanCharge, MonetaryCurrency currency);

Comment (adamsaghy) [reply fineract-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/mapper/LoanChargeMapper.java:41]: If i changed to return HashSet, it generated a dummy method:
```
    @Override
    public HashSet<LoanChargeCommand> map(Set<LoanCharge> loanCharge, MonetaryCurrency currency) {
        if ( loanCharge == null && currency == null ) {
            return null;
        }

        HashSet<LoanChargeCommand> hashSet = new HashSet<LoanChargeCommand>();

        return hashSet;
    }

**Review by alberto-art3ch — APPROVED** [review]
In genral LGTM, Thanks!

**Comment by SavvyPabi** [comment]
Hi @adamsaghy @vidakovic 

Great work on this restructure! Now that the foundational cleanup for loan creation/modification is merged, I wanted to share a Product Perspective on the logical next steps for the validation layer.

To build on this new structure, I suggest we look at standardizing our Business Validation & Rejection logic:

Standardized Reason Codes: Moving from generic strings to specific codes, so front-end apps can provide clearer user guidance.

Audit Persistence: Ensuring that "Rejected" or "Draft" attempts are logged in the audit trail rather than discarded—a major requirement for banking compliance.

Tiered Validation: Implementing a "fail-fast" approach (Technical/Schema checks first, then Business Policy) to optimize performance for high-volume environments.

I'm happy to help document a set of Standardized Business Reason Codes that align with international microfinance standards if that helps the post-refactor roadmap!

**Pull request merged**